rust/flexbuffers/src/builder/mod.rs

```diff
@@ -0,0 +1,370 @@
+            store_value(&mut self.buffer, x, child_width);
+            self.values.push(
+                Value::Reference {
+                    address,
+                    child_width,
+                    fxb_type: FlexBufferType::$Indirect,
+                }
+            );
+        }
+    }
+}
+#[derive(Debug, Clone)]
+struct CachedKey {
+    address: usize,
+    length: usize,
+}
+/// **Use this struct to build a Flexbuffer.**
+///
+/// Flexbuffers may only have a single root value, which may be constructed
+/// with  one of the following functions.
+/// * `build_singleton` will push 1 value to the buffer and serialize it as the root.
+/// * `build_vector` returns a `VectorBuilder`, into which many (potentially
+/// heterogenous) values can be pushed. The vector itself is the root and is serialized
+/// when the `VectorBuilder` is dropped (or `end` is called).
+/// * `build_map` returns a `MapBuilder`, which is similar to a `VectorBuilder` except
+/// every value must be pushed with an associated key. The map is serialized when the
+/// `MapBuilder` is dropped (or `end` is called).
+///
+/// These functions reset and overwrite the Builder which means, while there are no
+/// active `MapBuilder` or `VectorBuilder`, the internal buffer is empty or contains a
+/// finished Flexbuffer. The internal buffer is accessed with `view`.
+///
+/// The default builder interns keys. In the future, there should be string interning,
+/// and potentially key-vector interning.
+// TODO(cneo): Stablize options before 1.0.
+#[derive(Debug, Clone)]
+pub struct Builder {
+    buffer: Vec<u8>,
+    values: Vec<Value>,
+    key_pool: Option<Vec<CachedKey>>,
+}
+impl Default for Builder {
+    fn default() -> Self {
+        Builder {
+            key_pool: Some(Vec::new()),
+            values: Vec::new(),
+            buffer: Vec::new(),
+        }
+    }
+}
+
+
+impl<'a> Builder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+    /// Shows the internal flexbuffer. It will be empty on initialization or if `reset` was called.
+    /// Otherwise after `build_map`, `build_vector`, or `build_singleton` the internal flexbuffer
+    /// will be completed.
+    pub fn view(&self) -> &[u8] {
+        &self.buffer
+    }
+    /// Resets the internal state. Automatically called before building a new flexbuffer.
+    pub fn reset(&mut self) {
+        self.buffer.truncate(0);
+        self.values.truncate(0);
+        if let Some(pool) = self.key_pool.as_mut() {
+            pool.truncate(0);
+        }
+    }
+    fn push_key(&mut self, key: &str) {
+        assert!(
+            key.bytes().all(|b| b != b'\0'),
+            "Keys must not have internal nulls."
+        );
+        // Search key pool if there is one.
+        let found = self.key_pool.as_ref().map(|pool| {
+            pool.binary_search_by(|ck| {
+                let old_key = unsafe { read_trusted_str(&self.buffer, ck.address, ck.length) };
```

Comment:
Can you remove the use of `unsafe` by treating the data as &[u8] instead of str?

Reply:
yea I think so, something analogous to lazy_strcmp in src/reader/map.rs can get rid of the unsafe

Reply:
If &[u8] works here then that solves multiple problems.

Reply:
btw, this appears to do a binary_search always. At least in C++, while default on, this is optional. For example when serializing an object with a single root map, it be a total waste to do these searches.

Reply:
This does the binary search iff there's a key pool which I think matches C++.

Granted,  I've haven't implemented a way to construct a builder without a key pool (`BuilderFlag` in C++) nor string pooling nor key_vector pooling

Reply:
Added BuilderOptions which is analogous to BuilderFlag in C++. Setting SHARE_NONE means the binary search won't be called